src/components/CanvasWrapper.tsx

```diff
@@ -0,0 +1,183 @@
+
+import { ColorRamp, ColorizeCanvas, Dataset, Track } from "../colorizer";
+import { DrawMode } from "../colorizer/ColorizeCanvas";
+
+export type DrawSettings = {
+  mode: DrawMode;
+  color: Color;
+};
+
+type CanvasWrapperProps = {
+  canv: ColorizeCanvas;
+  /** Dataset to look up track and ID information in.
+   * Changing this does NOT update the canvas dataset; do so
+   * directly by calling `canv.setDataset()`.
+   */
+  dataset: Dataset | null;
+  showTrackPath: boolean;
+  outOfRangeDrawSettings: DrawSettings;
+  outlierDrawSettings: DrawSettings;
+  colorRamp: ColorRamp;
+  colorRampMin: number;
+  colorRampMax: number;
+  selectedTrack: Track | null;
+
+  /** Called when the mouse hovers over the canvas; reports the currently hovered id. */
+  onMouseHover?: (id: number) => void;
+  /** Called when the mouse exits the canvas. */
+  onMouseLeave?: () => void;
+  /** Called when the canvas is clicked; reports the track info of the clicked object. */
+  onTrackClicked?: (track: Track | null) => void;
+
+  maxWidth?: number;
+  maxHeight?: number;
+};
+
+const defaultProps: Partial<CanvasWrapperProps> = {
+  onMouseHover() {},
+  onMouseLeave() {},
+  onTrackClicked: () => {},
+  maxWidth: 730,
+  maxHeight: 500,
+};
+
+/**
+ * Provides a React component-style interface for interacting with ColorizeCanvas.
+ *
+ * Note that some canvas operations (like `setFrame`, `setFeature`, `setDataset`)
+ * are async and should be called directly on the canvas instance.
+ */
+export default function CanvasWrapper(inputProps: CanvasWrapperProps): ReactElement {
+  const props = { ...defaultProps, ...inputProps } as Required<CanvasWrapperProps>;
+
+  const canv = props.canv;
+  const canvasRef = useRef<HTMLDivElement>(null);
+  const isMouseOverCanvas = useRef(false);
+  const lastMousePositionPx = useRef([0, 0]);
+
+  // CANVAS PROPERTIES /////////////////////////////////////////////////
+
+  // Mount the canvas to the wrapper's location in the document.
+  useEffect(() => {
+    canvasRef.current?.parentNode?.replaceChild(canv.domElement, canvasRef.current);
+  }, []);
+
+  // Update canvas color ramp
+  useMemo(() => {
+    canv.setColorRamp(props.colorRamp);
+    canv.setColorMapRangeMin(props.colorRampMin);
+    canv.setColorMapRangeMax(props.colorRampMax);
+  }, [props.colorRamp, props.colorRampMin, props.colorRampMax]);
+
+  // Update drawing modes for outliers + out of range values
+  useMemo(() => {
+    const settings = props.outOfRangeDrawSettings;
+    canv.setOutOfRangeDrawMode(settings.mode, settings.color);
+  }, [props.outOfRangeDrawSettings]);
+  useMemo(() => {
+    const settings = props.outlierDrawSettings;
+    canv.setOutlierDrawMode(settings.mode, settings.color);
+  }, [props.outlierDrawSettings]);
+
+  // Updated track-related settings
+  useMemo(() => {
+    canv.setSelectedTrack(props.selectedTrack);
+    canv.setShowTrackPath(props.showTrackPath);
+  }, [props.selectedTrack, props.showTrackPath]);
```

Comment:
I'd again prefer these to be `useEffect`s, though if there's a good lifecycle reason not to do that (e.g. you found from testing that `useEffect` happens too late for interactions to feel nice) then this is an acceptable approach.

Reply:
Because `canv` breaks the React lifecycle model, these updates need to happen as part of the render or else React won't redraw the canvas (and then things like changing the color ramp will happen one update too late ;-;).

Reply:
Maybe you could add a comment saying why this should be a useMemo?